Add IP Address Validation and Separate IP and URI Logic to Files

DavidTPate opened the pull request:
This PR adds the ability to do IP address validation and updates the handling of the components used to build Regular Expressions for both IP and URI validation into files. Additionally, this replaces the old usage of `Array.reduce()` which was occurring before in both IP and URI validation construction in favor of the standard `for(...)` loop.

This resolves #607 and #617

DavidTPate commented:
@Marsup I got them split out where I thought they should be. I left the validation of options within `String` still since that is where I think it belongs, I can move it if desired to the appropriate file.

Marsup commented on lib/language.js:
```diff
@@ -116,6 +116,9 @@ exports.errors = {
         creditCard: 'must be a credit card',
-        ref: 'references "{{ref}}" which is not a number'
+        ref: 'references "{{ref}}" which is not a number',
+        ip: 'must be a valid ip address',
```
Regroup those under one option.

Marsup commented on lib/string.js:
```diff
@@ -374,8 +201,70 @@ internals.String.prototype.email = function (isEmailOptions) {
+                ipOptions.version = [ipOptions.version];
+            }
+
+            var version;
```
You don't need that outside the loop.

Marsup commented on lib/string.js:
```diff
@@ -388,27 +277,28 @@ internals.String.prototype.uri = function (uriOptions) {
                 uriOptions.scheme = [uriOptions.scheme];
             }
 
+            var scheme;
```
Inside loop as well.

Marsup commented on lib/string.js:
```diff
@@ -388,27 +277,28 @@ internals.String.prototype.uri = function (uriOptions) {
 
                 // Add OR separators if a value already exists
-                accumulator = accumulator ? accumulator + '|' : '';
+                customScheme = customScheme ? customScheme + '|' : '';
```
+= is just as good and clearer imho.

Marsup commented on lib/string.js:
```diff
@@ -388,27 +277,28 @@ internals.String.prototype.uri = function (uriOptions) {
                 // If someone wants to match HTTP or HTTPS for example then we need to support both RegExp and String so we don't escape their pattern unknowingly.
                 if (scheme instanceof RegExp) {
-                    return accumulator + scheme.source;
+                    customScheme = customScheme + scheme.source;
```
+= as well.

Marsup commented on lib/string.js:
```diff
@@ -388,27 +277,28 @@ internals.String.prototype.uri = function (uriOptions) {
+                    customScheme = customScheme + scheme.source;
+                } else {
+                    Hoek.assert(/[a-zA-Z][a-zA-Z0-9+-\.]*/.test(scheme), 'scheme at position ' + i + ' must be a valid scheme');
+                    customScheme = customScheme + Hoek.escapeRegex(scheme);
```
And also here.

Marsup commented on lib/string.js:
```diff
@@ -374,8 +201,70 @@ internals.String.prototype.email = function (isEmailOptions) {
+    }
+
+    return this._test('ip', ipOptions, function (value, state, options) {
+        if (regex.test(value)) {
```
Line break.

Marsup commented on lib/string.js:
```diff
@@ -374,8 +201,70 @@ internals.String.prototype.email = function (isEmailOptions) {
 
 
+internals.String.prototype.ip = function (ipOptions) {
+    var regex = internals.ipRegex;
```
Line break.

Marsup commented on lib/string.js:
```diff
@@ -374,8 +201,70 @@ internals.String.prototype.email = function (isEmailOptions) {
+            if (!ipOptions.version && ipOptions.cidr !== 'optional') {
+                regex = Ip.createIpRegex(['ipv4', 'ipv6', 'ipvfuture'], ipOptions.cidr);
+            }
+        } else {
```
Line break.

Marsup commented on lib/string/ip.js:
```diff
@@ -0,0 +1,33 @@
+    var regex;
+    var version;
+
+    for (var i = 0, il = versions.length; i < il; i++) {
```
Pre-increment for coherence.

Marsup commented on lib/string/ip.js:
```diff
@@ -0,0 +1,33 @@
+};
+
+internals.Ip.createIpRegex = function(versions, cidr) {
+    var regex;
```
Line break.

Marsup commented on lib/string/ip.js:
```diff
@@ -0,0 +1,33 @@
+
+internals.Ip.createIpRegex = function(versions, cidr) {
+    var regex;
+    var version;
```
Inside the loop.

Marsup commented on lib/string.js:
```diff
@@ -374,8 +201,70 @@ internals.String.prototype.email = function (isEmailOptions) {
+        }
+
+        if (ipOptions.version) {
+            if (!Array.isArray(ipOptions.version)) {
```
You should check the length if it's an array as well, createIpRegex will be invalid otherwise.

DavidTPate replied:
Added the check (and a test) for the situation for both uri and ip validation.

Marsup commented:
You're missing many line breaks in the tests as well but I won't comment on all of them.

DavidTPate commented:
Gotcha. I'll go through on the pieces you commented on, as well as the rest of the code. I've been getting more familiar with the coding style from the `contrib` repo.

DavidTPate commented:
@Marsup I think it's all set now. I went through and tried to find all the syntax issues I could and addressed your feedback. I also figured out what I did wrong when trying to squash things before (I wasn't force pushing before), so I squashed the changes together.

Marsup commented on lib/string.js:
```diff
@@ -374,8 +201,68 @@ internals.String.prototype.email = function (isEmailOptions) {
+
+    var regex = internals.ipRegex;
+    ipOptions = ipOptions || {};
+    Hoek.assert(typeof ipOptions === 'object', 'ipOptions must be an object');
```
simply options

Marsup commented on lib/string.js:
```diff
@@ -374,8 +201,68 @@ internals.String.prototype.email = function (isEmailOptions) {
+    Hoek.assert(typeof ipOptions === 'object', 'ipOptions must be an object');
+
+    if (ipOptions.cidr) {
+        Hoek.assert(typeof ipOptions.cidr === 'string', 'ipOptions.cidr must be a string');
```
only cidr

Marsup commented on lib/string.js:
```diff
@@ -374,8 +201,68 @@ internals.String.prototype.email = function (isEmailOptions) {
+        Hoek.assert(typeof ipOptions.cidr === 'string', 'ipOptions.cidr must be a string');
+        ipOptions.cidr = ipOptions.cidr.toLowerCase();
+
+        Hoek.assert(ipOptions.cidr in Ip.cidrs, 'ipOptions.cidr must be one of ' + Object.keys(Ip.cidrs).join(', '));
```
same

Marsup commented on lib/string.js:
```diff
@@ -374,8 +201,68 @@ internals.String.prototype.email = function (isEmailOptions) {
+            ipOptions.version = [ipOptions.version];
+        }
+
+        Hoek.assert(ipOptions.version.length >= 1, 'ipOptions.version must have at least 1 version specified');
```
only version

Marsup commented on lib/string.js:
```diff
@@ -374,8 +201,68 @@ internals.String.prototype.email = function (isEmailOptions) {
+        var versions = [];
+        for (var i = 0, il = ipOptions.version.length; i < il; ++i) {
+            var version = ipOptions.version[i];
+            Hoek.assert(typeof version === 'string', 'ipOptions.version at position ' + i + ' must be a string');
```
I think you're getting the idea from now :)

Marsup commented on lib/string.js:
```diff
@@ -388,27 +275,30 @@ internals.String.prototype.uri = function (uriOptions) {
 
-            // Flatten the array into a string to be used to match the schemes.
-            customScheme = uriOptions.scheme.reduce(function(accumulator, scheme, index) {
+            Hoek.assert(uriOptions.scheme.length >= 1, 'uriOptions.scheme must have at least 1 scheme specified');
```
Same as on IP

Marsup commented on test/string.js:
```diff
@@ -627,6 +627,515 @@ describe('string', function () {
     });
 
+    describe('#ip', function () {
+        var invalidIPs = [
```
Line break

Marsup commented on test/string.js:
```diff
@@ -627,6 +627,515 @@ describe('string', function () {
+        });
+
+        describe('#ip({ version: [ "ipv4", "ipv6" ] })', function() {
+            it('should validate all ipv4 and ipv6 addresses with a default CIDR strategy', function(done) {
```
Line break

Marsup commented on test/string.js:
```diff
@@ -627,6 +627,515 @@ describe('string', function () {
+        });
+
+        describe('#ip({ version: "ipvfuture" })', function() {
+            it('should validate all ipvfuture addresses with a default CIDR strategy', function(done) {
```
Line break

Marsup commented:
Do you know how to rebase your commits or should I do it ?

DavidTPate commented:
@Marsup It should be all squashed and include the changes for your comments about the error messages.

Marsup commented:
Your patch contains things that have nothing to do with this, you must have failed the rebase.

DavidTPate commented:
Darn! Must have been that merge previously to catch up to the 6.1.0 branch. I'll take a look.

Marsup commented:
@DavidTPate where are we on this ?

DavidTPate commented:
@Marsup The changes only include those for the PR now. That was really odd, I had changed permissions and items that I didn't expect included in my commits from when I merged somehow. Never had that happen before.

Marsup commented:
Rebased on branch, thanks.
